File: apps/web/src/state/watchOfficer.ts
import { useCallback, useEffect, useState } from 'react';
import { apiFetch } from '../transport/http.js';

export interface WatchOfficerEvidence {
  domain: string;
  severity: string;
  summary: string;
  lon: number;
  lat: number;
  ref?: string;
  kind?: 'measured' | 'inferred';
  basis?: string;
}

export interface WatchOfficerBrief {
  id: string;
  key: string;
  created: number;
  threat_level: 'high' | 'elevated' | 'low' | string;
  domains: string[];
  centroid: { lon?: number; lat?: number };
  title: string;
  narrative?: string;
  evidence: WatchOfficerEvidence[];
  follow_up: string[];
  playbook: Record<string, unknown>;
  status: string;
}

export interface WatchOfficerPlaybook {
  id: string;
  trigger: string;
  action: string;
}

export interface WatchOfficerStatus {
  running: boolean;
  cycle_s: number;
  sweeps: number;
  open: number;
  by_level: Record<string, number>;
  total_filed: number;
  last_sweep_at: number | null;
  last_filed_at: number | null;
  playbooks: WatchOfficerPlaybook[];
}

const POLL_MS = 30_000;

/** Poll the watch-officer live telemetry — proves the autonomous loop is alive
 *  and shows what it will DO (playbooks). Null until the first response. */
export function useWatchOfficerStatus(): WatchOfficerStatus | null {
  const [status, setStatus] = useState<WatchOfficerStatus | null>(null);
  useEffect(() => {
    let alive = true;
    const load = async (): Promise<void> => {
      try {
        const r = await apiFetch('/api/watch-officer/status');
        if (!r.ok) return;
        const j = (await r.json()) as WatchOfficerStatus;
        if (alive) setStatus(j);
      } catch {
        /* transient */
      }
    };
    void load();
    const t = setInterval(() => void load(), POLL_MS);
    return () => {
      alive = false;
      clearInterval(t);
    };
  }, []);
  return status;
}

/** Poll the watch-officer draft briefs + expose dismiss/ack. Optimistically
 *  drops a brief locally on triage so the UI feels instant; the next poll
 *  reconciles with the backend. */
export function useWatchOfficerBriefs(): {
  briefs: WatchOfficerBrief[];
  dismiss: (id: string) => void;
  ack: (id: string) => void;
} {
  const [briefs, setBriefs] = useState<WatchOfficerBrief[]>([]);

  useEffect(() => {
    let alive = true;
    const load = async (): Promise<void> => {
      try {
        const r = await apiFetch('/api/watch-officer/briefs');
        if (!r.ok) return;
        const j = (await r.json()) as { briefs?: WatchOfficerBrief[] };
        if (alive) setBriefs(j.briefs ?? []);
      } catch {
        /* transient — next tick retries */
      }
    };
    void load();
    const t = setInterval(() => void load(), POLL_MS);
    return () => {
      alive = false;
      clearInterval(t);
    };
  }, []);

  const triage = useCallback((id: string, verb: 'dismiss' | 'ack'): void => {
    setBriefs((bs) => bs.filter((b) => b.id !== id));
    void apiFetch(`/api/watch-officer/briefs/${encodeURIComponent(id)}/${verb}`, { method: 'POST' });
  }, []);

  const dismiss = useCallback((id: string) => triage(id, 'dismiss'), [triage]);
  const ack = useCallback((id: string) => triage(id, 'ack'), [triage]);

  return { briefs, dismiss, ack };
}

export interface BriefElaboration {
  ok: boolean;
  id: string;
  text: string;
  model?: string;
  backend?: string;
  cached?: boolean;
}

/** Fetch a deeper AI write-up of one brief (grounded in its evidence). Returns
 *  null on any failure — the caller shows the brief without elaboration. 409
 *  means selection inference is off; surfaced as a typed sentinel so the UI can
 *  prompt the operator to enable it rather than silently doing nothing. */
export async function elaborateBrief(id: string): Promise<BriefElaboration | 'disabled' | null> {
  try {
    const r = await apiFetch(`/api/watch-officer/briefs/${encodeURIComponent(id)}/elaborate`, {
      method: 'POST',
    });
    if (r.status === 409) return 'disabled';
    if (!r.ok) return null;
    return (await r.json()) as BriefElaboration;
  } catch {
    return null;
  }
}

File: apps/api/app/routes/watch_officer.py
"""Watch-officer briefs — /api/watch-officer/*.

Read + triage the draft briefs the ``intel.watch_officer`` loop files. Keyless,
like ``/api/alerts`` — this is a personal analyst surface reading in-process state,
and gating it behind ``current_user`` would 401 whenever Supabase is unset (the
standing-detections trap). The briefs are derived from public fusion output; there
is nothing user-scoped to protect here.
"""

from __future__ import annotations

import json
from typing import Any

from fastapi import APIRouter, Depends, HTTPException

from app import llm, upstream
from app.intel import watch_officer
from app.keys import UserCtx, current_user_or_local

router = APIRouter(tags=["watch-officer"])

# Deeper AI write-up of a filed brief. Cached per brief id — briefs are immutable
# once filed, so an elaboration never needs recomputing (a re-file after escalation
# gets a fresh id, so it re-elaborates). Longer than the selection-brief TTL since
# the input can't drift.
_ELABORATE_TTL_S = 900.0
_ELABORATE_MAX_TOKENS = 512


@router.get("/api/watch-officer/status")
async def wo_status() -> dict[str, Any]:
    """Live telemetry — is the officer running, its cadence, sweep/brief counts,
    and the playbooks it can fire. Lets the UI prove the agent is alive."""
    return watch_officer.status()


@router.get("/api/watch-officer/briefs")
async def briefs() -> dict[str, Any]:
    """Open draft briefs the watch-officer has filed, newest first."""
    return {"briefs": watch_officer.list_briefs()}


@router.post("/api/watch-officer/briefs/{bid}/dismiss")
async def dismiss_brief(bid: str) -> dict[str, Any]:
    """Drop a brief as noise. 404 if unknown/expired."""
    if not watch_officer.dismiss(bid):
        raise HTTPException(status_code=404, detail="unknown brief")
    return {"ok": True, "id": bid}


@router.post("/api/watch-officer/briefs/{bid}/ack")
async def ack_brief(bid: str) -> dict[str, Any]:
    """Acknowledge a brief (operator saw the finding). 404 if unknown/expired."""
    if not watch_officer.ack(bid):
        raise HTTPException(status_code=404, detail="unknown brief")
    return {"ok": True, "id": bid}


def _elaborate_prompt(brief: dict[str, Any]) -> tuple[str, str]:
    """Grounded system+user prompt for a deeper analytic write-up of one brief."""
    system = (
        "You are a senior OSINT watch analyst. You are handed a watch-officer "
        "brief: a fused, cited convergence of live signals the automated loop "
        "flagged. Write 3-5 sentences of markdown that go DEEPER than the "
        "one-line narrative: (1) what this convergence most likely represents; "
        "(2) why — grounded in the specific evidence rows (cite the ids, "
        "distances, gaps); (3) what would confirm or refute it; (4) the single "
        "most useful next action. Ground every claim in the brief — never "
        "invent ids, positions, or events, and never overstate intent. End "
        "with a bold **Confidence: low | medium | high** reflecting how much "
        "the evidence supports the read."
    )
    payload = {
        "title": brief.get("title"),
        "threat_level": brief.get("threat_level"),
        "domains": brief.get("domains"),
        "centroid": brief.get("centroid"),
        "narrative": brief.get("narrative"),
        "evidence": brief.get("evidence"),
        "follow_up": brief.get("follow_up"),
        "playbook": brief.get("playbook"),
    }
    user = json.dumps(payload, default=str, separators=(",", ":"))[:4000]
    return system, user


@router.post("/api/watch-officer/briefs/{bid}/elaborate")
async def elaborate_brief(
    bid: str, _ctx: UserCtx = Depends(current_user_or_local)
) -> dict[str, Any]:
    """Deeper AI assessment of one brief, grounded in its evidence chain. Cached
    per brief id. 404 if the brief is unknown/expired; 409 if selection inference
    is disabled (same gate as the selection brief); 502 if the model fails."""
    brief = watch_officer.get_brief(bid)
    if brief is None:
        raise HTTPException(status_code=404, detail="unknown brief")
    if not llm.selection_enabled():
        raise HTTPException(status_code=409, detail="selection inference is disabled")

    cache_key = f"wo-elaborate:{bid}"
    computed = False

    async def _load() -> dict[str, Any]:
        nonlocal computed
        computed = True
        system, user = _elaborate_prompt(brief)
        res = await llm.chat(
            [{"role": "system", "content": system}, {"role": "user", "content": user}],
            tier="selection",
            max_tokens=_ELABORATE_MAX_TOKENS,
            label="watch_officer.elaborate",
        )
        if not res.ok:
            raise HTTPException(status_code=502, detail=res.error or "elaboration failed")
        return {"ok": True, "id": bid, "text": res.text, "model": res.model, "backend": res.backend}

    payload = await upstream.cache.get_or_fetch(cache_key, _ELABORATE_TTL_S, _load)
    return {**payload, "cached": not computed}
